In [13]:
#Task 1.1
import pandas as pd
import numpy as np

df = pd.read_csv('dataset_ifn580_262_v1.csv')

columns_to_check = [
    'destination',
    'passanger',
    'weather',
    'coupon',
    'expiration',
    'gender',
    'maritalStatus',
    'education',
    'occupation'
]

for col in columns_to_check:
    print(col, ":", df[col].unique())

destination : ['No Urgent Place' 'Home' 'Work']
passanger : ['Alone' 'Friend(s)' 'Kid(s)' 'Partner']
weather : ['Sunny' 'Rainy' 'Snowy']
coupon : ['Restaurant(<20)' 'Coffee House' 'Carry out & Take away' 'Bar'
 'Restaurant(20-50)']
expiration : ['1d' '2h']
gender : ['Female' 'Male']
maritalStatus : ['Unmarried partner' 'Single' 'Married partner' 'Divorced' 'Widowed']
education : ['Some college - no degree' 'Bachelors degree' 'Associates degree'
 'High School Graduate' 'Graduate degree (Masters or Doctorate)'
 'Some High School']
occupation : ['Unemployed' 'Architecture & Engineering' 'Student'
 'Education&Training&Library' 'Healthcare Support'
 'Healthcare Practitioners & Technical' 'Sales & Related' 'Management'
 'Arts Design Entertainment Sports & Media' 'Computer & Mathematical'
 'Life Physical Social Science' 'Personal Care & Service'
 'Community & Social Services' 'Office & Administrative Support'
 'Construction & Extraction' 'Legal' 'Retired'
 'Installation Maintenance & Repair' 

# Missing Value

- 'car' has approximately 99.15% missing values and may be unreliable for modelling.
- 'temperature' has a small proportion of missing values.
- 'Bar', 'CoffeeHouse', 'CarryAway', 'RestaurantLessThan20', and 'Restaurant20To50' also have small proportions of missing values.
- The restaurant/visit-frequency variables contain ordered categories, so they should be treated as ordinal rather than nominal.

'Temperture' float64/ 1.970987 [55. 80. nan 30.]

'Bar' object/ 0.843582 

'Coffeehouse' object/ 1.710817

'Carryaway' object/ 1.190476

'RestaurantLessThan20' object/ 1.024913

'Restaurant20To50' object/ 1.490066

# Data exploration

Age:
[21, 46, 26, 31, 41, 50plus, 36, below21]
- object in pandas
- contains numeric categories plus textual boundaries such as below21 / 50plus
- ordered meaning
- likely ordinal

income:
[$37500-$49999, $62500-$74999, $12500-$24999, $75000-$87499, $50000-$62499, $25000-$37499, $100000 or more, $87500-$99999, less than $12500]
- object in pandas
- income range represented as text
- ordered meaning
- likely ordinal

time:
[2PM, 10AM, 6PM, 7AM, 10PM]
- object in pandas
- clock time represented as text with AM/PM
- requires numerical/ordered representation before modelling

In [14]:
# Task 1.2
missing_count = df.isnull().sum()
missing_percent = df.isnull().mean() * 100

missing_summary = pd.DataFrame({
    'Missing Count': missing_count,
    'Missing Percentage': missing_percent
})

missing_summary = missing_summary[missing_summary['Missing Count'] > 0]

display(missing_summary)

,Missing Count,Missing Percentage
temperature,250,1.970987
car,12576,99.148534
Bar,107,0.843582
CoffeeHouse,217,1.710817
CarryAway,151,1.190476
RestaurantLessThan20,130,1.024913
Restaurant20To50,189,1.490066


Car:
- missing rate> 99%
- too little observed information
- Impuation would be unreliable
- Remove the variable

In [15]:
print("Mean:", df['temperature'].mean())
print("Median:", df['temperature'].median())

Mean: 63.28976998552356
Median: 80.0


Temperature:
- has 250 missing values, which is a relatively small proportion of the dataset.
- The mean is 63.289, while the median is 80.
- The original temperature values are 30, 55, and 80.
- Median imputation is selected because it keeps the imputed value within the existing temperature values and avoids introducing a new value that does not appear in the original dataset.

In [17]:
for col in ['Bar']:
    print(col)
    print(df[col].value_counts(dropna=False))
    print()

Bar
Bar
never    5197
less1    3482
1~3      2473
4~8      1076
gt8       349
NaN       107
Name: count, dtype: int64



In [ ]:
Bar:
- has 107 missing values, which is a small proportion of the dataset.
- The variable contains ordered frequency categories such as never, less1, 1~3, 4~8, and gt8.
- The most frequent category is never
- Mode imputation is selected because the variable is categorical and the missing proportion is low.

In [18]:
for col in ['CoffeeHouse']:
    print(col)
    print(df[col].value_counts(dropna=False))
    print()

CoffeeHouse
CoffeeHouse
less1    3385
1~3      3225
never    2962
4~8      1784
gt8      1111
NaN       217
Name: count, dtype: int64



In [ ]:
CoffeeHouse:
- has 217 missing values, which is a small proportion of the dataset.
- The variable contains ordered frequency categories such as never, less1, 1~3, 4~8, and gt8.
- The most frequent category is less1
- Mode imputation is selected because the variable is categorical and the missing proportion is low.

In [19]:
for col in ['CarryAway']:
    print(col)
    print(df[col].value_counts(dropna=False))
    print()

CarryAway
CarryAway
1~3      4672
4~8      4258
less1    1856
gt8      1594
never     153
NaN       151
Name: count, dtype: int64



CarryAway:
- has 151 missing values, which is a small proportion of the dataset.
- The variable contains ordered frequency categories such as never, less1, 1~3, 4~8, and gt8.
- The most frequent category is 1~3.
- Mode imputation is selected because the variable is categorical and the missing proportion is low.

In [20]:
for col in ['RestaurantLessThan20']:
    print(col)
    print(df[col].value_counts(dropna=False))
    print()

RestaurantLessThan20
RestaurantLessThan20
1~3      5376
4~8      3580
less1    2093
gt8      1285
never     220
NaN       130
Name: count, dtype: int64



RestaurantLessThan20:
- has 130 missing values, which is a small proportion of the dataset.
- The variable contains ordered frequency categories such as never, less1, 1~3, 4~8, and gt8.
- The most frequent category is 1~3.
- Mode imputation is selected because the variable is categorical and the missing proportion is low.

In [21]:
for col in ['Restaurant20To50']:
    print(col)
    print(df[col].value_counts(dropna=False))
    print()

Restaurant20To50
Restaurant20To50
less1    6077
1~3      3290
never    2136
4~8       728
gt8       264
NaN       189
Name: count, dtype: int64



Restaurant20To50:
- has 189 missing values, which is a small proportion of the dataset.
- The variable contains ordered frequency categories such as never, less1, 1~3, 4~8, and gt8.
- The most frequent category is less1.
- Mode imputation is selected because the variable is categorical and the missing proportion is low.

In [22]:
# Task 1.3 missing-value preprocessing
# missing-value preprocessing
df = df.drop(columns=['car'])
df['temperature'] = df['temperature'].fillna(df['temperature'].median())
df['Bar'] = df['Bar'].fillna(df['Bar'].mode()[0])
df['CoffeeHouse'] = df['CoffeeHouse'].fillna(df['CoffeeHouse'].mode()[0])
df['CarryAway'] = df['CarryAway'].fillna(df['CarryAway'].mode()[0])
df['RestaurantLessThan20'] = df['RestaurantLessThan20'].fillna(df['RestaurantLessThan20'].mode()[0])
df['Restaurant20To50'] = df['Restaurant20To50'].fillna(df['Restaurant20To50'].mode()[0])

df.isnull().sum()

destination             0
passanger               0
weather                 0
temperature             0
time                    0
coupon                  0
expiration              0
gender                  0
age                     0
maritalStatus           0
has_children            0
education               0
occupation              0
income                  0
Bar                     0
CoffeeHouse             0
CarryAway               0
RestaurantLessThan20    0
Restaurant20To50        0
toCoupon_GEQ5min        0
toCoupon_GEQ15min       0
toCoupon_GEQ25min       0
direction_same          0
direction_opp           0
is_accepted             0
dtype: int64

In [ ]:
# encoding


In [23]:
df['age'].unique()

array(['21', '46', '26', '31', '41', '50plus', '36', 'below21'],
      dtype=object)